### PR TITLE
ADR-19: Python Web Framework

### PR DESCRIPTION
**Pull request by snakedye**
An ADR detailing the decision process of the next web framework used in the [fertiscan-backend](https://github.com/ai-cfia/fertiscan-backend/).

https://github.com/ai-cfia/fertiscan-backend/issues/148

### CODE REVIEW

File: adr/019-python-web-framework.en-ca.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+## Executive Summary
+
+This section provides a brief summary of the decision being made, outlining the
+main points so that the reader can understand the gist without reading the
+entire document.
+
+## Context
+
+Currently, we use Flask to build our APIs. While Flask has been sufficient, it is synchronous and lacks some modern features like built-in asynchronous support and automatic API documentation (OpenAPI/Swagger). There are other open-source frameworks that could better suit our needs by offering enhanced developer experience and feature sets. We are considering alternatives like FastAPI, Quart, and others that meet our criteria, but only open-source solutions will be considered.
```
Comment (k-allagbe) [inline adr/019-python-web-framework.en-ca.md:25]: I suggest [rewrap](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) to fix line length lint error.

**Comment by k-allagbe** [comment]
What's your recommendation?

**Comment by snakedye** [comment]
@k-allagbe FastAPI for the built-in documentation. I want to experiment with it first before giving a definitive approval.

File: adr/019-python-web-framework.en-ca.md (hunk 1 above)
Comment (snakedye) [reply adr/019-python-web-framework.en-ca.md]: I use rewrap but it doesn't factor in the rules of this repo.

Comment (k-allagbe) [reply adr/019-python-web-framework.en-ca.md]: I think you can configure the length. Let me try.

**Comment by ChromaticPanic** [comment]
How does this ADR impact other projects? Since Nachet is currently using Quart, choosing FastAPI means the team will have to support both frameworks, or we would have to convert Nachet. 

I disagree with some of the cons for Quart, a lot of Flask extensions and libraries work with Quart it is maintained by the same team. 

You can also get swagger documentation in flask https://github.com/flasgger/flasgger

Performance wise sure FastAPI is more performant, one question is, is the expected load on our backend high enough to merit switching from Quart to FastAPI?

I'd like to see a consideration in this document on how much work it would take to convert our projects to FastAPI, converting Flask to Quart is very simple.

**Comment by snakedye** [comment]
@ChromaticPanic 
> How does this ADR impact other projects? Since Nachet is currently using Quart, choosing FastAPI means the team will have to support both frameworks, or we would have to convert Nachet.

Initially, the scope was on FertiScan, but you're raising a good point. Supporting another framework would be a burden on the entire team regardless.

> You can also get swagger documentation in flask https://github.com/flasgger/flasgger

That's what we do, but it's tedious to manually change it every time we update the API. Sometimes we forget to correct some details. With FastAPI, it's all automatic.

> I'd like to see a consideration in this document on how much work it would take to convert our projects to FastAPI, converting Flask to Quart is very simple.

https://github.com/ai-cfia/fertiscan-backend/tree/148-fast-api-port

Thanks for the feedback. I'll make the necessary corrections in the ADR and raise the points you mentioned.

**Comment by snakedye** [comment]
@ChromaticPanic We'd like to have your opinion regarding the framework to choose. I made some changes to the ADR to reflect our recent findings and your suggestions.

**Comment by ChromaticPanic** [comment]
> @ChromaticPanic We'd like to have your opinion regarding the framework to choose. I made some changes to the ADR to reflect our recent findings and your suggestions.

Thanks I'll take a better look once my accounts are setup! Hopefully everything goes smoothly tomorrow.

File: adr/019-python-web-framework.en-ca.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+   easily, providing flexibility.
+
+**Cons :**
+
+1. **Learning Curve**: While FastAPI is developer-friendly, beginners might find
+   the reliance on type hints and async a bit challenging.
+2. **Smaller Ecosystem**: Though growing, FastAPI’s ecosystem is not as large or
+   mature as Flask’s. Fewer extensions and third-party libraries are available
+   compared to Flask.
+3. **Less Mature**: As a newer framework, it may not be as battle-tested in
```
[the remark this answers is not in the document]
Comment (k-allagbe) [reply adr/019-python-web-framework.en-ca.md:104]: Sentence still seems incomplete.

**Review by k-allagbe — APPROVED** [review]

**Review by ChromaticPanic — APPROVED** [review]

**Pull request merged**